### PR TITLE
bpo-37860: re-add netlify.toml to set up deploy previews for docs

### LINKED ISSUE #82041: Add deploy preview for docs
BPO | [37860](https://bugs.python.org/issue37860)
--- | :---
Nosy | @willingc, @ewdurbin, @Mariatta, @isidentical, @epicfaace, @aeros
PRs | <li>python/cpython#15288</li><li>python/cpython#18242</li><li>python/cpython#19025</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2019-08-14.21:14:46.493>
labels = ['3.9', 'docs']
title = 'Add deploy preview for docs'
updated_at = <Date 2021-03-22.01:23:26.366>
user = 'https://github.com/epicfaace'
```

bugs.python.org fields:
```python
activity = <Date 2021-03-22.01:23:26.366>
actor = 'epicfaace'
assignee = 'docs@python'
closed = False
closed_date = None
closer = None
components = ['Documentation']
creation = <Date 2019-08-14.21:14:46.493>
creator = 'epicfaace'
dependencies = []
files = []
hgrepos = []
issue_num = 37860
keywords = ['patch']
message_count = 7.0
messages = ['349752', '350084', '364264', '364266', '364271', '368201', '389269']
nosy_count = 7.0
nosy_names = ['docs@python', 'willingc', 'EWDurbin', 'Mariatta', 'BTaskaya', 'epicfaace', 'aeros']
pr_nums = ['15288', '18242', '19025']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue37860'
versions = ['Python 3.9']
```

</p></details>

### PR DESCRIPTION
**Pull request by epicfaace**
Fixes https://github.com/python/cpython/issues/82041 by setting up Netlify to set up deploy previews for docs.

This PR re-does https://github.com/python/cpython/pull/15288. Reverts python/cpython#30272 (where we had originally removed netlify.toml due to build limits). As per https://github.com/python/cpython/issues/82041#issuecomment-1120452334, the build limits are now large enough that they should be fine to use for cpython.

### CODE REVIEW

**Comment by epicfaace** [comment]
Do we need a NEWS label for this? I was thinking we could skip news?

**Comment by hugovk** [comment]
Good work! Out of interest, how long does a single build/deploy take?

**Comment by epicfaace** [comment]
It took 4m16s.

**Comment by hugovk** [comment]
Thanks, so with 25,000 minutes per month, 4-5 minutes per build gives us 5,000-6,250 builds per month.

That's significantly better than the ~170/month estimate 2 years ago: https://github.com/python/cpython/pull/15288#issuecomment-579476340.

I see it's also possible to configure ignored builds, can we only build previews when there are changes in the `Doc` directory? That would save resources and make PR runs quicker/cleaner too.

https://docs.netlify.com/configure-builds/ignore-builds/

**Comment by epicfaace** [comment]
> I see it's also possible to configure ignored builds, can we only build previews when there are changes in the `Doc` directory? That would save resources and make PR runs quicker/cleaner too.
> 
> https://docs.netlify.com/configure-builds/ignore-builds/

This is already configured by default. The link above says "Depending on your [Branches settings](https://docs.netlify.com/site-deploys/overview/#branch-deploy-controls), any time there is a change in a linked repository, Netlify tries to determine whether there are changes in the site’s [base directory](https://docs.netlify.com/configure-builds/overview/#definitions-1) by comparing the last known version of the files in that directory."

The `base` directory is already set to `Doc`.

**Comment by hugovk** [comment]
Great, then I think we're all set!

---

@ewdurbin As a follow on to https://github.com/python/cpython/pull/15288#issuecomment-524120287 in 2019, we should now have 25k minutes per month (https://github.com/python/cpython/issues/82041#issuecomment-1120452334), and are ready to try this out again.

Please could you hook up the Netlify account? I would suggest we only run it on `main` to begin, and review usage after a few weeks/a month, and consider enabling it for `3.11` too (possibly even others).

---

If we have max 3 team members (https://github.com/python/cpython/pull/15288#issuecomment-524120287), @Mariatta, you asked to be added in 2019 (https://github.com/python/cpython/pull/15288#issuecomment-524120287), would you still like to?

And to further increase the bus factor, who would like to be the 3rd?

**Comment by epicfaace** [comment]
I'm happy to be the third! If it's fine though I'm not a committer

**Comment by epicfaace** [comment]
Also, looks like we're no longer limited to 3 team members max: see https://www.netlify.com/legal/open-source-policy/:

> Our Open Source plan has the same features and limits as our [Pro plan](https://www.netlify.com/pricing/#teams), with the addition of free unlimited team members.

**Comment by epicfaace** [comment]
Also, one more thing: I think to comply with the requirements mentioned in https://www.netlify.com/legal/open-source-policy/, we should add a Netlify badge / "This site is powered by Netlify" message on all deploy previews.

I don't think we need to add it to the main docs, but we can add it on deploy previews only -- for example, we could set an environment variable that then renders the Netlify badge only on deploy previews. I'm happy to implement if we're good with that approach. What do you think @hugovk ?

**Comment by hugovk** [comment]
The env var approach sounds good to add it only to deploy previews. Perhaps the text link option, in the footer? And just on the homepage or all pages?

**Comment by Mariatta** [comment]
Thanks for your work @epicface.

I think we need to display more prominently that the docs rendered is a "preview". So where-ever we put the netlify badge, we should add the "preview" info as well. Is that something we can do?

As comparison, on the devguide (built with readthedocs), the preview is displayed at the top of the page:

e.g: https://cpython-devguide--867.org.readthedocs.build/

Screen Shot 2022-05-18 at 10 30 25 AM

**Comment by AA-Turner** [comment]
Adding a warning admonition to the top of the page seems a reasonable idea regardless, although the ephemeral nature of the previews (& non-canonical URL) should help to disambiguate d.p.o from the previews. 

@hugovk from the linked Netifly page, it seems only the "main page" is required - so `/index.html`.

A

**Comment by hugovk** [comment]
Hi again @epicfaace! Do you think you can make the above changes? Thanks!

**Comment by arhadthedev** [comment]
I think the popup warning that collapses into the bottom will be more noticeable. Here is an [example](https://ci.tc39.es/preview/tc39/ecma262/sha/f2d38813f7f29f851525e71af37a3e2637656b71/) from tc39/ecma262:



image

image

**Comment by epicfaace** [comment]
> Hi again @epicfaace! Do you think you can make the above changes? Thanks!

Yes, doing so!

@arhadthedev I like that idea. relevant code here: https://github.com/tc39/ecma262/blob/267fdbdc716677bc9934906a498cf9bad4bfb70d/scripts/snapshot_warning.html https://github.com/tc39/ecma262/blob/267fdbdc716677bc9934906a498cf9bad4bfb70d/scripts/insert_snapshot_warning.js and https://github.com/tc39/ecma262/blob/267fdbdc716677bc9934906a498cf9bad4bfb70d/scripts/snapshot_warning.css

**Comment by epicfaace** [comment]
All right, this PR should be ready now. I just added a simple banner for now:

image

Sample deploy preview: https://62af84a185fcf70008d07339--steady-heliotrope-731cac.netlify.app/

File: Doc/tools/templates/layout.html [hunk 1]
```diff
@@ -8,6 +8,19 @@
...
+{%- if is_deployment_preview %}
+<div id="deployment-preview-warning" style="padding: .5em; text-align: center; background-color: #fff2ba; color: #6a580e;">
+  <div style="float: right; margin-top: -10px; margin-left: 10px;">
+    <a href="https://www.netlify.com">
+      <img src="https://www.netlify.com/img/global/badges/netlify-color-accent.svg" alt="Deploys by Netlify" />
+    </a>
+  </div>
+  {% trans %}This is a deployment preview that was created from a <a href="{{ repository_url }}/pull/{{ pr_id }}">pull request</a>.
+  For authoritative documentation, see the {% endtrans %}
+  <a href="/3/{{ pagename }}{{ file_suffix }}">{% trans %} Python documentation for the current stable release{% endtrans %}</a>.
```
Comment (hugovk) [inline Doc/tools/templates/layout.html]: Let's make this a bit more consice:
```suggestion
  {% trans %}This is a deploy preview created from a <a href="{{ repository_url }}/pull/{{ pr_id }}">pull request</a>.
  For authoritative documentation, see the {% endtrans %}
  <a href="/3/{{ pagename }}{{ file_suffix }}">{% trans %} the current stable release{% endtrans %}</a>.
```

Comment (hugovk) [inline Doc/tools/templates/layout.html]: Your [demo](https://62af84a185fcf70008d07339--steady-heliotrope-731cac.netlify.app/) links to [here](https://62af84a185fcf70008d07339--steady-heliotrope-731cac.netlify.app/3/index.html) which is a 404.

I guess the merged one will also have Netlify URL. Should this URL be an absolute link to `https://docs.python.org/3/` ?

**Review by hugovk — COMMENTED** [review]
Thanks, a couple of comments!

File: Doc/tools/templates/layout.html (hunk 1 above)
Comment (epicfaace) [reply Doc/tools/templates/layout.html]: Done

Comment (epicfaace) [reply Doc/tools/templates/layout.html]: Good catch -- just updated

**Comment by hugovk** [comment]
Om mobile, the top of the banner is hidden:

image

**Comment by epicfaace** [comment]
> Om mobile, the top of the banner is hidden:

Note that this banner uses the same styling as the "This document is for an old version of Python that is no longer supported" banner. The top of this banner will also be hidden once we start showing this banner on versions of the docs that have a responsive mobile layout. This means that when 3.8 reaches EOL (as 3.8 is the first version of the docs with a responsive mobile layout), we'll have a similar problem in which the top of the banner "This document is for an old version of Python that is no longer supported." will be hidden.

**Comment by epicfaace** [comment]
to clarify -- 3.6 doesn't have this problem as it's not using a mobile responsive layout:

image

but 3.8 most certainly will, once we add the red banner.

image

**Comment by hugovk** [comment]
Ah okay. So for the "no longer supported" it won't happen until 3.8 is EOL on 2024-10-14, but for the previews it will happen right away :)

We could report this to (I guess) https://github.com/python/python-docs-theme and come up with a fix later?

**Comment by epicfaace** [comment]
@hugovk yes. It seems more involved to fix that issue so I've went ahead and reported it in https://github.com/python/python-docs-theme/issues/95.

**Review by hugovk — APPROVED** [review]
Thanks!

**Comment by hugovk** [comment]
@ewdurbin Hi! Please could you hook up the Netlify account again?

See https://github.com/python/cpython/pull/92852#issuecomment-1129292036

And please add @Mariatta to the team. Thank you!

**Comment by encukou** [comment]
@ewdurbin Will you be able to do this, or is something blocking you?

**Comment by hugovk** [comment]
Looks like we're all set now!

Here's a recent docs PR build:

* https://github.com/python/cpython/pull/92181

And the corresponding preview:

* And https://deploy-preview-92181--python-cpython-preview.netlify.app/

Thank you Ee for setting up Netlify and Ashwin for this and the earlier PRs! :clap:

**Comment by hugovk** [comment]
I guess because that PR doesn't yet have the banner changes from this one. 

@encukou Would you like to rebase/merge https://github.com/python/cpython/pull/100054 from `main` to test it?

**Comment by Mariatta** [comment]
This is amazing! Thank you everyone for pushing through and thanks @epicfaace for this PR. This is a great help to all contributors here.

**Comment by epicfaace** [comment]
Amazing, thank you all for reviewing and getting this in!

**Pull request merged**